### PR TITLE
[*] Connected the alter entry use case to the GUI and added support for splitter bill.

### PR DESCRIPTION
**Pull request by ScottCTD**
# Summary

- I completely added support for alter entry use case and splitter bill.

# Tasks
- [x] [*]Connected the alter entry use case with the GUI.
- [x] [+] Added a `ConstraintTableCellEditor` to constrain the value that a user can input to a table cell. 
- [x] [*] Fixed some bugs in `MySQLDatabaseGateway`. @Lei-Tin 
- [x] [*] Cleaned up some convention issues and optimized some imports in some classes. 
- [x] [*] Added a global variable `Main.DATETIME_PATTERN`.
- [x] [*] Fixed the bug that the splitter bill columns do not change after switching.
- [x] [*] Fixed that the back button is enabled even in the normal bill.
- [x] [*] Added support of splitter bill in `AlterEntryUseCase` and `DeleteEntryUseCase`.
- [x] [*] Fixed the bug related to chaning the columns of the bill table.

### CODE REVIEW

File: src/main/java/billgates/database/MySQLDatabaseGateway.java [hunk 1]
```diff
@@ -506,7 +507,7 @@ public void modifyEntry(int billId, int entryId, String column, String newValue)
             // If equal, we are modifying the entry from the user's bill
             query = String.format("""
                         UPDATE bill_%d
-                        SET %s = "%s"
+                        SET `%s` = '%s'
```
Comment (Lei-Tin) [inline src/main/java/billgates/database/MySQLDatabaseGateway.java:510]: Does adding ticks (`) like this not affect how we insert the other types? 

Also, good catch with the double quotes " and single quotes '. I googled and it seems like single quote is more preferred.

File: src/test/java/database/MySQLDatabaseGatewayTests.java [hunk 2]
```diff
@@ -1,8 +1,13 @@
 package database;
 
-import org.junit.*;
-import static org.junit.Assert.*;
-import billgates.database.*;
+import billgates.Main;
+import billgates.database.MySQLDatabaseGateway;
+import billgates.database.QueryBillData;
+import billgates.database.QueryEntryData;
```
Comment (Lei-Tin) [inline src/test/java/database/MySQLDatabaseGatewayTests.java:6]: With the pull request I attempted [here](https://github.com/CSC207-2022F-UofT/course-project-bill-gates/pull/76)

I believe we should remove the classes `QueryBillData` and `QueryEntryData`. 
We can fix that when solving conflicts in merges

**Review by Lei-Tin — APPROVED** [review]
Everything looks great!

**Review by Yuhan-Ut — COMMENTED** [review]
Everything looks great here! Seems you successfully connect the alter entry uses case into the main and related files!  
But I have a small question here, seems you did not use the alter entry input port in Main, is that necessary to import this interface? or this is just for convention?

**Review by Yuhan-Ut — APPROVED** [review]

**Comment by ScottCTD** [comment]
> But I have a small question here, seems you did not use the alter entry input port in Main, is that necessary to import this interface? or this is just for convention?

I actually used `AlterEntryInputPort`.

File: src/main/java/billgates/database/MySQLDatabaseGateway.java (hunk 1 above)
Comment (ScottCTD) [reply src/main/java/billgates/database/MySQLDatabaseGateway.java:510]: For keywords like `from`, if you do not quote them using ```, then you will get an error.

File: src/main/java/billgates/use_cases/user_join/UserJoinUseCase.java [hunk 3]
```diff
@@ -36,7 +36,8 @@ public UserJoinUseCase(DatabaseGateway gateway, UserJoinOutputPort outputPort) {
     @Override
     public void join(UserJoinRequestModel model) {
         List<QueryUserData> usersData = this.gateway.getUserData();
```
Comment (Lei-Tin) [inline src/main/java/billgates/use_cases/user_join/UserJoinUseCase.java:38]: Actually I just realized, we can just test if `this.gateway.getUserData(username)` returns `null` or not. 

Remember we made this method return `null` whenever we cannot find the specified username?

**Review by Lei-Tin — APPROVED** [review]
GOOD REFACTORS

I think we can just merge this

**Pull request merged**